### PR TITLE
Level set: composition (II)

### PR DESCRIPTION
**Pull request by peterrum**
Follow up PR to #14.

### CODE REVIEW

File: include/adaflo/util.h [hunk 1]
```diff
@@ -52,4 +54,78 @@ locally_owned_subdomain(const MeshType &mesh)
   return tria_parallel != nullptr ? tria_parallel->locally_owned_subdomain() : 0;
 }
 
+template <int dim>
+void
+compute_cell_diameters(const MatrixFree<dim, double> &         matrix_free,
+                       const unsigned int                      dof_index,
+                       AlignedVector<VectorizedArray<double>> &cell_diameters,
+                       double &                                cell_diameter_min,
+                       double &                                cell_diameter_max)
```
Comment (peterrum) [inline include/adaflo/util.h:63]: @mschreter FYI :)

**Comment by peterrum** [comment]
@mschreter Note that I have rebased to incorporate your changes.

File: source/level_set_okz_preconditioner.cc [hunk 2]
```diff
@@ -178,6 +178,17 @@ initialize_projection_matrix(
     ilu_projection_matrix.initialize(projection_matrix);
   }
 }
+template void
+initialize_projection_matrix<1, double, VectorizedArray<double>>(
+  const MatrixFree<1, double, VectorizedArray<double>> &matrix_free,
```
Comment (peterrum) [inline source/level_set_okz_preconditioner.cc:183]: Thx!

File: source/level_set_okz_reinitialization.cc [hunk 3]
```diff
@@ -346,5 +364,6 @@ LevelSetOKZSolverReinitialization<dim>::reinitialize(const double       dt,
   this->time_stepping.next();
 }
 
+template class LevelSetOKZSolverReinitialization<1>;
```
Comment (peterrum) [inline source/level_set_okz_reinitialization.cc:367]: FYI @mschreter

File: include/adaflo/level_set_okz_compute_curvature.h [hunk 4]
```diff
@@ -76,7 +75,6 @@ class LevelSetOKZSolverComputeCurvature
 {
 public:
   LevelSetOKZSolverComputeCurvature(
-    LevelSetOKZSolverComputeNormal<dim> &                  normal_operator,
```
Comment (peterrum) [inline include/adaflo/level_set_okz_compute_curvature.h:79]: The curvature operator does not need the normal operator anymore.

File: source/level_set_okz.cc [hunk 5]
```diff
@@ -462,8 +461,17 @@ template <int dim>
...
+  // compute the curvature directly from C, but that is less accurate. TODO:
+  // include that variant by a parameter
+  {
+    TimerOutput::Scope timer(*this->timer, "LS compute normal.");
+    this->compute_normal(false);
+  }
+  {
+    TimerOutput::Scope timer(*this->timer, "LS compute curvature.");
+    curvatur_operator->compute_curvature(diffuse_large_values);
+  }
```
Comment (peterrum) [inline source/level_set_okz.cc:474]: ... instead we need to call the normal operator ourselves.

**Review by peterrum — COMMENTED** [review]
FYI @mschreter

File: include/adaflo/level_set_okz_reinitialization.h [hunk 6]
```diff
@@ -68,7 +68,6 @@ class LevelSetOKZSolverReinitialization
   using BlockVectorType = LinearAlgebra::distributed::BlockVector<double>;
 
   LevelSetOKZSolverReinitialization(
-    LevelSetOKZSolverComputeNormal<dim> &             normal_operator,
```
Comment (peterrum) [inline include/adaflo/level_set_okz_reinitialization.h:71]: No need to pass the normal operator to the constructor.

File: include/adaflo/level_set_okz_reinitialization.h [hunk 7]
```diff
@@ -103,10 +101,10 @@ class LevelSetOKZSolverReinitialization
...
   // performs reinitialization
   virtual void
-  reinitialize(const double       dt,
-               const unsigned int stab_steps,
-               const unsigned int diff_steps                              = 0,
-               const bool         diffuse_cells_with_large_curvature_only = false);
+  reinitialize(const double                     dt,
+               const unsigned int               stab_steps,
+               const unsigned int               diff_steps     = 0,
+               const std::function<void(bool)> &compute_normal = [](const bool) {});
```
Comment (peterrum) [inline include/adaflo/level_set_okz_reinitialization.h:107]: Instead a lambda with the effect.

**Review by peterrum — COMMENTED** [review]
@mschreter I have tried also to eliminate the normal operator from the reinitialization operator. It is significantly more complicated...

File: source/level_set_okz.cc (hunk 5 above)
Comment (mschreter) [reply source/level_set_okz.cc:473]: I will follow the same strategy in meltpool.

**Comment by peterrum** [comment]
@kronbichler This PR should be good to be merged. We have adjusted the interface so that they can be better be used from external. Furthermore, we have enabled that parts of the code can be compiled for 1D.

**Review by kronbichler — APPROVED** [review]

**Pull request merged**